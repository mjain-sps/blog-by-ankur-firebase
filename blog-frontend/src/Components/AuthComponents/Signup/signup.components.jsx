import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";

import {
  AuthContainer,
  AuthTitle,
  AuthInput,
  AuthForm,
  AuthFormControl,
  AuthLabel,
} from "../signin.styles";

import ButtonComponent from "../../Button/button.component";
import Messages from "../../Notifications/messages.component";
import LoaderComponent from "../../Loader/loader.component";
//importing Actions
import { signUpAction } from "../../../Actions/auth.actions";

//Main component starts
const Signup = (props) => {
  const { loading, user, error } = useSelector((state) => state.user);

  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [password2, setPassword2] = useState("");

  const dispatch = useDispatch();

  useEffect(() => {
    if (user) {
      props.history.push("/");
    }
  }, [props.history, user]);

  const handleSignup = async (e) => {
    e.preventDefault();
    if (!email || !password2 || !password) {
      alert("Cannot leave fields blank. Please fill form completely");
    } else if (password !== password2) {
      alert("password do not match");
    } else {
      const userData = { email, password };
      dispatch(signUpAction(userData));
    }
  };

  return loading ? (
    <LoaderComponent />
  ) : (
    <AuthContainer>
      {error && <Messages>{error}</Messages>}

      <AuthTitle>Signup</AuthTitle>

      <AuthForm>
        <AuthFormControl>
          <AuthInput
            type="email"
            id="email"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
          />
          <AuthLabel htmlFor="email">Enter Email</AuthLabel>
        </AuthFormControl>

        <AuthFormControl>
          <AuthInput
            type="password"
            id="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
          <AuthLabel>Password</AuthLabel>
        </AuthFormControl>

        <AuthFormControl>
          <AuthInput
            type="password"
            id="password2"
            value={password2}
            onChange={(e) => setPassword2(e.target.value)}
          />
          <AuthLabel>Password Again</AuthLabel>
        </AuthFormControl>

        <ButtonComponent
          type="submit"
          onClick={handleSignup}
          theme="primary"
          width="50"
        >
          Login
        </ButtonComponent>
      </AuthForm>
    </AuthContainer>
  );
};

export default Signup;
