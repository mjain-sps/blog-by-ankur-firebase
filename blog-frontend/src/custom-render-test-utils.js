import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import rootReducer from "./Reducers";

const customRender = (
  ui,
  {
    initialState,
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    ...options
  } = {}
) => {
  const WrapperWithRoutes = ({ children }) => {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  };
  return render(ui, { wrapper: WrapperWithRoutes, ...options });
};

export * from "@testing-library/react";
export { customRender };
