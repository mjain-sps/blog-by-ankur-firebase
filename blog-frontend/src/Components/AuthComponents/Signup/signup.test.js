import React from "react";
import { customRender, fireEvent } from "../../../custom-render-test-utils";
import "@testing-library/jest-dom/extend-expect";
import Signup from "./signup.components";

describe("Signup components tests", () => {
  test("To take a snapshot of component", () => {
    const { container } = customRender(<Signup />);
    expect(container).toMatchSnapshot();
  });

  test("check input email", () => {
    const { getByTestId } = customRender(<Signup />);
    const emailInput = getByTestId("email");
    fireEvent.change(emailInput, { target: { value: "madhur" } });
    expect(emailInput.value).toBe("madhur");
  });

  test("check input password", () => {
    const { getByTestId } = customRender(<Signup />);
    const passwordInput = getByTestId("password");
    fireEvent.change(passwordInput, { target: { value: "XYZ" } });
    expect(passwordInput.value).toBe("XYZ");
  });

  test("check input password2", () => {
    const { getByTestId } = customRender(<Signup />);
    const passwordInput2 = getByTestId("password2");
    fireEvent.change(passwordInput2, { target: { value: "Ramjane" } });
    expect(passwordInput2.value).toBe("Ramjane");
  });

  test("check button test if all inputs are found to be correct", (done) => {
    const { getByTestId, getByText } = customRender(<Signup />);
    const emailInput = getByTestId("email");
    const passwordInput = getByTestId("password");
    const passwordInput2 = getByTestId("password2");
    fireEvent.change(emailInput, { target: { value: "madhur" } });
    fireEvent.change(passwordInput, { target: { value: "XYZ" } });
    fireEvent.change(passwordInput2, { target: { value: "XYZ" } });
    const buttonElement = getByText(/login/i);
    expect.assertions(1);
    fireEvent.click(buttonElement);
    done();
    expect(window.location.href).toBe("http://localhost/");
  });

  test("check button test if email is left blank", (done) => {
    jest.spyOn(window, "alert").mockImplementation(() => {});

    const { getByTestId, getByText } = customRender(<Signup />);
    // const emailInput = getByTestId("email");
    const passwordInput = getByTestId("password");
    const passwordInput2 = getByTestId("password2");
    // fireEvent.change(emailInput, { target: { value: "" } });
    fireEvent.change(passwordInput, { target: { value: "XYZ" } });
    fireEvent.change(passwordInput2, { target: { value: "XYZ" } });
    const buttonElement = getByText(/login/i);

    fireEvent.click(buttonElement);
    done();
    expect(window.alert).toHaveBeenCalledTimes(1);
  });

  test("check button test if password is left blank", (done) => {
    jest.spyOn(window, "alert").mockImplementation(() => {});

    const { getByTestId, getByText } = customRender(<Signup />);
    const emailInput = getByTestId("email");
    // const passwordInput = getByTestId("password");
    const passwordInput2 = getByTestId("password2");
    fireEvent.change(emailInput, { target: { value: "" } });
    // fireEvent.change(passwordInput, { target: { value: "XYZ" } });
    fireEvent.change(passwordInput2, { target: { value: "XYZ" } });
    const buttonElement = getByText(/login/i);

    fireEvent.click(buttonElement);
    done();
    expect(window.alert).toHaveBeenCalledTimes(1);
  });

  test("check button test if password2 is left blank", (done) => {
    jest.spyOn(window, "alert").mockImplementation(() => {});

    const { getByTestId, getByText } = customRender(<Signup />);
    const emailInput = getByTestId("email");
    const passwordInput = getByTestId("password");
    // const passwordInput2 = getByTestId("password2");
    fireEvent.change(emailInput, { target: { value: "" } });
    fireEvent.change(passwordInput, { target: { value: "XYZ" } });
    // fireEvent.change(passwordInput2, { target: { value: "XYZ" } });
    const buttonElement = getByText(/login/i);

    fireEvent.click(buttonElement);
    done();
    expect(window.alert).toHaveBeenCalledTimes(1);
  });

  test("check button test if password and password2 do not match", (done) => {
    jest.spyOn(window, "alert").mockImplementation(() => {});

    const { getByTestId, getByText } = customRender(<Signup />);
    const emailInput = getByTestId("email");
    const passwordInput = getByTestId("password");
    const passwordInput2 = getByTestId("password2");
    fireEvent.change(emailInput, { target: { value: "mjain.sps@gmail.com" } });
    fireEvent.change(passwordInput, { target: { value: "XYZ" } });
    fireEvent.change(passwordInput2, { target: { value: "XYAAZ" } });
    const buttonElement = getByText(/login/i);

    fireEvent.click(buttonElement);
    done();
    expect(window.alert).toHaveBeenCalledTimes(1);
  });

  test("to check what happens is we pass an initial state with user object", () => {
    const userObject = {
      id: "XYZ",
      name: "madhur jain",
      email: "mjain@gmail.com",
    };
    customRender(<Signup />, {
      initialState: { user: userObject },
    });
    expect(window.location.href).toBe("http://localhost/");
  });
});
