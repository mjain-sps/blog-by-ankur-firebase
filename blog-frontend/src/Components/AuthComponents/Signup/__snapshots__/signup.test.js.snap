// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup components tests To take a snapshot of component 1`] = `
<div>
  <div
    class="sc-bdnxRM dkmzgK"
  >
    <div
      class="sc-gtsrHT gAQZhd"
    >
      Signup
    </div>
    <form
      class="sc-dlnjwi hrJjEv"
    >
      <div
        class="sc-hKFxyN kodvPp"
      >
        <input
          class="sc-eCApnc bgDbDW"
          data-testid="email"
          id="email"
          type="email"
          value=""
        />
        <label
          class="sc-jSFjdj jigzJV"
          for="email"
        >
          Enter Email
        </label>
      </div>
      <div
        class="sc-hKFxyN kodvPp"
      >
        <input
          class="sc-eCApnc bgDbDW"
          data-testid="password"
          id="password"
          type="password"
          value=""
        />
        <label
          class="sc-jSFjdj jigzJV"
        >
          Password
        </label>
      </div>
      <div
        class="sc-hKFxyN kodvPp"
      >
        <input
          class="sc-eCApnc bgDbDW"
          data-testid="password2"
          id="password2"
          type="password"
          value=""
        />
        <label
          class="sc-jSFjdj jigzJV"
        >
          Password Again
        </label>
      </div>
      <button
        class="sc-gKAaRy fcfiWv"
        type="button"
        width="50"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
